osv/debian_version_cache.py

```diff
@@ -17,28 +17,42 @@
 from urllib import request
 import json
 
+from . import cache
+
 CLOUD_API_CACHE_URL = 'https://storage.googleapis.com/debian-osv/first_package_cache.json.gz'
-debian_version_cache = None
+CACHE_DURATION_SECONDS = 60 * 60 * 24
+
+debian_version_cache = cache.InMemoryCache()
+
+
+def _update_from_cloud_cache():
+  """Update the debian version cache from the Google cloud link"""
+  res = request.urlopen(CLOUD_API_CACHE_URL)
 
+  item: typing.Dict[str, typing.Dict] = json.loads(gzip.decompress(res.read()))
+  for release, sources in item.items():
+    if not sources:
+      continue
 
-def _initiate_from_cloud_cache(
-    force_reload: bool = False) -> typing.Dict[str, typing.Dict]:
-  """Load the debian version cache from the Google cloud link, if it hasn't been loaded yet"""
-  global debian_version_cache
-  if debian_version_cache is None or force_reload:
-    res = request.urlopen(CLOUD_API_CACHE_URL)
-    debian_version_cache = json.loads(gzip.decompress(res.read()))
+    for package, version in sources.items():
+      debian_version_cache.set((package, release), version,
+                               CACHE_DURATION_SECONDS)
 
 
 def get_first_package_version(package_name: str, release_number: str) -> str:
```

Comment:
Let's just create a Python decorator for these cases.

And let's restructure the caching like so:

e.g. 

```python
@cache.Cached(cache=debian_version_cache, ttl=CACHE_DURATION_SECONDS)
def get_debian_first_versions():
  ...

def get_first_package_version(...):
 first_versions = get_debian_first_versions()
 return first_versions.get((package_name, release_number))
```

cache.Cached can be implemented something similar to

```python
def Cached(cache: ..., ttl: ...)
  def decorator(func):
    @functools.wraps(func)
    def wrapper(*args, **kwargs):
       if cache_hasnt_expired:
         return cached_value
       
       value = func()
       set_cache(value)
       return value
    return wrapper
  
  return decorator
```

The key value can be constructed by doing some reflection (e.g. taking __file__ and the name of the function)

Reply:
This will change the the behavior of the cache, the current implementation will also try to update the cache when it can't find the package it is looking for. The proposed implementation from my understanding is essentially having a cache with one key, which is either expired or not expired.

Reply:
Actually, taking a step back here, we might not need to cache this at all. 

If instead, we structure our first packages version data like so in the bucket:

```json
gs://debian-osv/first_package_versions/<package>/<release>
```

Then we can just make a http request on-demand to get the latest without any risk of stale data.

Reply:
Good idea, added both the caching decorator and the new format, so each package can be retrieved individually, but since once received it is very unlikely to change, it's cached for 24 hours.